homeassistant/components/device_tracker/upc_connect.py

```diff
@@ -0,0 +1,164 @@
+        }
+
+        self.websession = async_create_clientsession(
+            hass, cookie_jar=aiohttp.CookieJar(unsafe=True, loop=hass.loop))
+
+    @asyncio.coroutine
+    def async_scan_devices(self):
+        """Scan for new devices and return a list with found device IDs."""
+        if self.token is None:
+            reconnect = yield from self.async_login()
+            if not reconnect:
+                _LOGGER.error("Not connected to %s", self.host)
+                return []
+
+        raw = yield from self._async_ws_function(CMD_DEVICES)
+        xml_root = ET.fromstring(raw)
+
+        return [mac.text for mac in xml_root.iter('MACAddr')]
+
+    @asyncio.coroutine
+    def async_get_device_name(self, device):
+        """The firmware doesn't save the name of the wireless device."""
+        return None
+
+    @asyncio.coroutine
+    def async_login(self):
+        """Login into firmware and get first token."""
+        response = None
+        try:
+            # get first token
+            with async_timeout.timeout(10, loop=self.hass.loop):
+                response = yield from self.websession.get(
+                    "http://{}/common_page/login.html".format(self.host)
+                )
+
+            self.token = self._async_get_token()
+
+            # login
+            data = yield from self._async_ws_function(CMD_LOGIN, {
+                'Username': 'NULL',
+                'Password': self.password,
+            })
+
+            # erfolgreich?
+            if data.find("successful") != -1:
+                return True
+            return False
+
+        except (asyncio.TimeoutError, aiohttp.errors.ClientError):
+            _LOGGER.error("Can not load loging page from %s", self.host)
+            return False
+
+        finally:
+            if response is not None:
+                yield from response.release()
+
+    @asyncio.coroutine
+    def _async_ws_function(self, function, additional_form=None):
+        """Execute a command on UPC firmware webservice."""
+        form_data = {
+            'token': self.token,
+            'fun': function
+        }
+
+        if additional_form:
+            form_data.update(additional_form)
+
+        response = None
+        try:
+            with async_timeout.timeout(10, loop=self.hass.loop):
+                response = yield from self.websession.post(
+                    "http://{}/xml/getter.xml".format(self.host),
+                    data=form_data,
+                    headers=self.headers
+                )
+
+                # error on UPC webservice
+                if response.status != 200:
+                    _LOGGER.warning(
+                        "Error %d on %s.", response.status, function)
+                    self.token = None
+                    return
+
+                # load data, store token for next request
+                raw = yield from response.text()
+                self.token = self._async_get_token()
```

Comment:
I would have thought cookies are in headers and only valid before the body of the response, but maybe it stays in the "cookiejar"

Do you always want to store the "token" or only on login?